File: lib/src/control/label.cpp
#include <sway/webui/control/label.h>
#include <sway/webcore/treeupdater.h>
#include <sway/webcore/css/stylesheet.h>

namespace sway::webui {

void Label::registerEmscriptenClass(lpcstr_t classname) {
	emscripten::class_<Label, emscripten::base<webcore::mvc::AView>>(classname)
		.constructor<core::NodePtr_t, std::string, webcore::TreeNodeElementCreateInfo>()
		.smart_ptr<LabelSmartPtr_t>("LabelSmartPtr_t")
		.class_function("create", &Label::create, emscripten::allow_raw_pointers())
		.function("setFontFamily", &Label::setFontFamily)
		.function("setColor", &Label::setColor)
		.function("getText", &Label::getText)
		.function("setText", &Label::setText);
}

LabelSmartPtr_t Label::create(core::NodePtr_t parent, const std::string & nodeId,
	const webcore::TreeNodeElementCreateInfo & createInfo, emscripten::val styleSheet, const std::string & content) {

	auto instance = std::make_shared<Label>(parent, nodeId, createInfo);
	instance->setStyleSheet(styleSheet);
	instance->setHtmlContent(content);
	parent->addChild(instance.get());
	return instance;
}

Label::Label(core::NodePtr_t parent,
	const std::string & nodeId, const webcore::TreeNodeElementCreateInfo & createInfo)
	: webcore::mvc::AView(parent, core::NodeIndex(), nodeId, createInfo) {
	// Empty
}

void Label::accept(webcore::ITreeVisitor * visitor) {
	visitor->visitOnEnter(this);

	for (core::NodePtr_t node : getChildren())
		static_cast<Label *>(node)->accept(visitor);
}

void Label::setFontFamily(const std::string & fontFamily) {
	_fontFamily = fontFamily;
}

void Label::setColor(const std::string & color) {
	_color = color;
}

std::string Label::getText() const {
	return getHtmlContent();
}

void Label::setText(const std::string & text) {
	if (text == getHtmlContent())
		return;

	setHtmlContent(text);
	for (core::HierarchyListener * listener : getHostTree()->getListeners())
		listener->onNodeUpdated(getNodeIndex());
}

} // namespace sway::webui

File: lib/src/control/label.ts
"use strict";

import * as _ from "underscore";

export class Label {
	public wasm_module: any = null;

	/**
	 * Конструктор представления.
	 * 
	 * @method constructor
	 * @param {any} wasm_context Контекст.
	 * @param {any} parent Родительский узел.
	 * @param {any} nodeId Уникальный идентификатор узла.
	 * @param {any} nodeCreateInfo Свойства представления.
	 * @param {any} stylesheet Стили.
	 * @param {String} text Текст.
	 */
	constructor(public wasm_context: any = null, parent: any,
		nodeId: string, nodeCreateInfo: any, stylesheet: any, text?: String) {

		this.wasm_module = wasm_context.module.Label.create(parent, nodeId,
			_.defaults(nodeCreateInfo, {
				tagname: "label"
			}), stylesheet, text);
	}
};
